File: java-samples/Advanced/secretsanta-cordapp/clients/src/main/java/net/corda/samples/secretsanta/webserver/Controller.java
package net.corda.samples.secretsanta.webserver;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import net.corda.samples.secretsanta.flows.CheckAssignedSantaFlow;
import net.corda.samples.secretsanta.flows.CreateSantaSessionFlow;
import net.corda.samples.secretsanta.states.SantaSessionState;
import net.corda.core.contracts.UniqueIdentifier;
import net.corda.core.identity.CordaX500Name;
import net.corda.core.identity.Party;
import net.corda.core.messaging.CordaRPCOps;
import net.corda.core.node.NodeInfo;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.io.IOException;
import java.util.*;

import com.sendgrid.*;
import com.sendgrid.helpers.mail.Mail;
import com.sendgrid.helpers.mail.objects.Content;
import com.sendgrid.helpers.mail.objects.Email;


/**
 * Define your API endpoints here.
 *
 * Note we allow all origins for convenience, this is NOT a good production practice.
 */
@RestController
@RequestMapping("/") // The paths for HTTP requests are relative to this base path.
@CrossOrigin(origins = "*")
public class Controller {
    private final CordaRPCOps proxy;
    private final CordaX500Name me;
    private final static Logger logger = LoggerFactory.getLogger(Controller.class);

    private boolean isMe(NodeInfo nodeInfo){
        return nodeInfo.getLegalIdentities().get(0).getName().equals(me);
    }

    private Party getPartyFromNodeInfo(NodeInfo nodeInfo) {
        Party target = nodeInfo.getLegalIdentities().get(0);
        return target;
    }

    public Controller(NodeRPCConnection rpc) {
        this.proxy = rpc.proxy;
        this.me = proxy.nodeInfo().getLegalIdentities().get(0).getName();
    }

    @RequestMapping(value = "/node", method = RequestMethod.GET)
    private ResponseEntity<String> returnName() {
        JsonObject resp = new JsonObject();
        resp.addProperty("name", me.toString());
        return ResponseEntity.status(HttpStatus.CREATED).contentType(MediaType.APPLICATION_JSON).body(resp.toString());
    }

    @RequestMapping(value = "/games/check", method = RequestMethod.POST)
    public ResponseEntity<String> checkGame(@RequestBody String payload) {

        System.out.println(payload);

        JsonObject convertedObject = new Gson().fromJson(payload, JsonObject.class);

        UniqueIdentifier gameId = UniqueIdentifier.Companion.fromString(convertedObject.get("gameId").getAsString());
        // NOTE lowercase the name for easy retrieve
        String playerName = "\"" + convertedObject.get("name").getAsString().toLowerCase().trim() + "\"";

        // optional param
        Boolean sendEmail = convertedObject.get("sendEmail").getAsBoolean();

        JsonObject resp = new JsonObject();

        try {
            SantaSessionState output = proxy.startTrackedFlowDynamic(CheckAssignedSantaFlow.class, gameId).getReturnValue().get();

            if (output.getAssignments().get(playerName) == null) {
                resp.addProperty("target", "target not found in this game");
                return ResponseEntity.status(HttpStatus.NOT_FOUND).contentType(MediaType.APPLICATION_JSON).body(resp.toString());
            }

            List<String> playerNames = output.getPlayerNames();
            List<String> playerEmails = output.getPlayerEmails();

            String playerEmail = playerEmails.get(playerNames.indexOf(playerName));
            String targetName = output.getAssignments().get(playerName).replace("\"", "");

            resp.addProperty("target", targetName);

            if (sendEmail) {
                System.out.println("sending email to "+ playerEmail);
                boolean b = sendEmail(playerEmail, craftNotice(playerName, targetName, gameId));

                if (!b) {
                    System.out.println("ERROR: Failed to send email.");
                }
            }

            return ResponseEntity.status(HttpStatus.ACCEPTED).contentType(MediaType.APPLICATION_JSON).body(resp.toString());
        } catch (Exception e) {
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.getMessage());
        }
    }

    @RequestMapping(value = "/games", method = RequestMethod.POST)
    public ResponseEntity<String> createGame(@RequestBody String payload) {

        System.out.println(payload);
        JsonObject convertedObject = new Gson().fromJson(payload, JsonObject.class);

        JsonArray pNames = convertedObject.getAsJsonArray("playerNames");
        JsonArray pMails = convertedObject.getAsJsonArray("playerEmails");

        // optional param
        Boolean sendEmail = convertedObject.get("sendEmail").getAsBoolean();

        List<String> playerNames = new ArrayList<>();
        List<String> playerEmails = new ArrayList<>();

        // NOTE: we lowercase all names internally for clarity
        for (JsonElement jo : pNames) {
            String newName = jo.toString().toLowerCase();

            if (!playerNames.contains(newName)) {
                playerNames.add(newName);
            }
        }
        for (JsonElement jo : pMails) {
            playerEmails.add(jo.toString());
        }

        try {
            Party elf = getPartyFromNodeInfo(proxy.networkMapSnapshot().get(2));

            // run the flow to create our game
            SantaSessionState output = proxy.startTrackedFlowDynamic(CreateSantaSessionFlow.class, playerNames, playerEmails, elf).getReturnValue().get().getTx().outputsOfType(SantaSessionState.class).get(0);
            UniqueIdentifier gameId = output.getLinearId();
            LinkedHashMap<String, String> assignments = output.getAssignments();

            System.out.println("Created Secret Santa Game ID# " + output.getLinearId().toString());

            // send email to each player with the assignments
            for (String p: playerNames) {
                String t = assignments.get(p).replace("\"", "");
                String msg = craftNotice(p, t, gameId);
                int ind = playerNames.indexOf(p);
                String email = playerEmails.get(ind).replace("\"", "");

                if (sendEmail) {
                    boolean b = sendEmail(email, msg);

                    if (!b) {
                        System.out.println("ERROR: Failed to send email.");
                    }
                }
            }

            JsonObject resp = new JsonObject();
            resp.addProperty("gameId", gameId.toString());
            return ResponseEntity.status(HttpStatus.CREATED).contentType(MediaType.APPLICATION_JSON).body(resp.toString());

        } catch (Exception e) {
            System.out.println("Exception : " + e.getMessage());
            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.getMessage());
        }
    }


    public String craftNotice(String p, String t, UniqueIdentifier gameId) {
        return "Hello, " + p + "!" + "\n" + "Your super secret santa target for game # " + gameId.toString() + " is " + t + ".";
    }

    public boolean sendEmail(String recipient, String msg) {

        // TODO replace the below line with your *REAL* api key!
        final String SENDGRID_API_KEY = "SG.xxxxxx_xxxxxxxxxxxxxxx.xxxxxxxxxxxxxxxx_xxxxxxxxxxxxxxxxxx_xxxxxxx";

        if (SENDGRID_API_KEY.equals("SG.xxxxxx_xxxxxxxxxxxxxxx.xxxxxxxxxxxxxxxx_xxxxxxxxxxxxxxxxxx_xxxxxxx")) {
            System.out.println("You haven't added your sendgrid api key!");
            return true;
        }

        // you'll need to specify your sendgrid verified sender identity for this to mail out.
        Email from = new Email("test@example.com");
        String subject = "Secret Assignment from the Elves";
        Email to = new Email(recipient);
        Content content = new Content("text/plain", msg);
        Mail mail = new Mail(from, subject, to, content);

        SendGrid sg = new SendGrid(SENDGRID_API_KEY);
        Request request = new Request();

        try {
            request.setMethod(Method.POST);
            request.setEndpoint("mail/send");
            request.setBody(mail.build());
            Response response = sg.api(request);
            System.out.println(response.getStatusCode());
            System.out.println(response.getBody());
            System.out.println(response.getHeaders());
            return true;
        } catch (IOException ex) {
            System.out.println(ex.toString());
            return false;
        }
    }
}

File: kotlin-samples/Advanced/secretsanta-cordapp/clients/src/main/kotlin/net/corda/samples/secretsanta/webserver/Controller.kt
package net.corda.samples.secretsanta.webserver

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.sendgrid.Method
import com.sendgrid.Request
import com.sendgrid.SendGrid
import com.sendgrid.helpers.mail.Mail
import com.sendgrid.helpers.mail.objects.Content
import com.sendgrid.helpers.mail.objects.Email
import net.corda.core.contracts.UniqueIdentifier
import net.corda.core.contracts.UniqueIdentifier.Companion.fromString
import net.corda.core.identity.Party
import net.corda.core.internal.toX500Name
import net.corda.core.node.NodeInfo
import net.corda.samples.secretsanta.flows.CheckAssignedSantaFlow
import net.corda.samples.secretsanta.flows.CreateSantaSessionFlow
import net.corda.samples.secretsanta.states.SantaSessionState
import org.bouncycastle.asn1.x500.X500Name
import org.bouncycastle.asn1.x500.style.BCStyle
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.MediaType.APPLICATION_JSON_VALUE
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.io.IOException
import java.util.*

/**
 * Define your API endpoints here.
 */
@RestController
@RequestMapping("/") // The paths for HTTP requests are relative to this base path.
@CrossOrigin(origins = ["*"])
class Controller(rpc: NodeRPCConnection) {

    private val proxy = rpc.proxy
    private val me = proxy.nodeInfo().legalIdentities.first().name

    companion object {
        private val logger = LoggerFactory.getLogger(RestController::class.java)
    }

    fun X500Name.toDisplayString() : String  = BCStyle.INSTANCE.toString(this)

    /** Helpers for filtering the network map cache. */
    private fun isNotary(nodeInfo: NodeInfo) = proxy.notaryIdentities().any { nodeInfo.isLegalIdentity(it) }
    private fun isMe(nodeInfo: NodeInfo) = nodeInfo.legalIdentities.first().name == me
    private fun isNetworkMap(nodeInfo: NodeInfo) = nodeInfo.legalIdentities.single().name.organisation == "Network Map Service"

    private fun getPartyFromNodeInfo(nodeInfo: NodeInfo): Party {
        return nodeInfo.legalIdentities[0]
    }

    /**
     * Returns the node's name.
     */
    @GetMapping(value = ["me"], produces = [APPLICATION_JSON_VALUE])
    fun whoami() = mapOf("me" to me.toString())

    /**
     * Returns all parties registered with the [NetworkMapService]. These names can be used to look up identities
     * using the [IdentityService].
     */
    @GetMapping(value = ["peers"], produces = [APPLICATION_JSON_VALUE])
    fun getPeers(): Map<String, List<String>> {
        return mapOf("peers" to proxy.networkMapSnapshot()
                .filter { isNotary(it).not() && isMe(it).not() && isNetworkMap(it).not() }
                .map { it.legalIdentities.first().name.toX500Name().toDisplayString() })
    }


    @RequestMapping(value = ["/node"], method = [RequestMethod.GET])
    private fun returnName(): ResponseEntity<String> {
        val resp = JsonObject()
        resp.addProperty("name", me.toString())
        return ResponseEntity.status(HttpStatus.CREATED).contentType(MediaType.APPLICATION_JSON).body(resp.toString())
    }


    @RequestMapping(value = ["/games/check"], method = [RequestMethod.POST])
    fun checkGame(@RequestBody payload: String?): ResponseEntity<String?>? {
        println(payload)
        val convertedObject = Gson().fromJson(payload, JsonObject::class.java)
        val gameId = fromString(convertedObject["gameId"].asString)
        // NOTE lowercase the name for easy retrieve
        val playerName = "\"" + convertedObject["name"].asString.toLowerCase().trim { it <= ' ' } + "\""

        // optional param
        val sendEmail = convertedObject["sendEmail"].asBoolean
        val resp = JsonObject()
        return try {
            val output = proxy.startTrackedFlowDynamic(CheckAssignedSantaFlow::class.java, gameId).returnValue.get()
            if (output.getAssignments()!![playerName] == null) {
                resp.addProperty("target", "target not found in this game")
                return ResponseEntity.status(HttpStatus.NOT_FOUND).contentType(MediaType.APPLICATION_JSON).body(resp.toString())
            }
            val playerNames = output.playerNames
            val playerEmails = output.playerEmails
            val playerEmail = playerEmails[playerNames.indexOf(playerName)]
            val targetName = output.getAssignments()!![playerName]!!.replace("\"", "")
            resp.addProperty("target", targetName)
            if (sendEmail) {
                println("sending email to $playerEmail")
                val b: Boolean = sendEmail(playerEmail, craftNotice(playerName, targetName, gameId))
                if (!b) {
                    println("ERROR: Failed to send email.")
                }
            }
            ResponseEntity.status(HttpStatus.ACCEPTED).contentType(MediaType.APPLICATION_JSON).body(resp.toString())
        } catch (e: Exception) {
            ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.message)
        }
    }

    @RequestMapping(value = ["/games"], method = [RequestMethod.POST])
    fun createGame(@RequestBody payload: String?): ResponseEntity<String?>? {
        println(payload)
        val convertedObject = Gson().fromJson(payload, JsonObject::class.java)
        val pNames = convertedObject.getAsJsonArray("playerNames")
        val pMails = convertedObject.getAsJsonArray("playerEmails")

        // optional param
        val sendEmail = convertedObject["sendEmail"].asBoolean
        val playerNames: MutableList<String> = ArrayList()
        val playerEmails: MutableList<String> = ArrayList()

        // NOTE: we lowercase all names internally for clarity
        for (jo in pNames) {
            val newName = jo.toString().toLowerCase()
            if (!playerNames.contains(newName)) {
                playerNames.add(newName)
            }
        }
        for (jo in pMails) {
            playerEmails.add(jo.toString())
        }
        return try {
            val elf: Party = getPartyFromNodeInfo(proxy.networkMapSnapshot()[2])

            // run the flow to create our game
            val output = proxy.startTrackedFlowDynamic(CreateSantaSessionFlow::class.java, playerNames, playerEmails, elf).returnValue.get().tx.outputsOfType(SantaSessionState::class.java)[0]
            val gameId = output.linearId
            val assignments = output.getAssignments()
            println("Created Secret Santa Game ID# " + output.linearId.toString())

            // send email to each player with the assignments
            for (p in playerNames) {
                val t = assignments!![p]!!.replace("\"", "")
                val msg: String = craftNotice(p, t, gameId)
                val ind = playerNames.indexOf(p)
                val email = playerEmails[ind].replace("\"", "")
                if (sendEmail) {
                    val b: Boolean = sendEmail(email, msg)
                    if (!b) {
                        println("ERROR: Failed to send email.")
                    }
                }
            }
            val resp = JsonObject()
            resp.addProperty("gameId", gameId.toString())
            ResponseEntity.status(HttpStatus.CREATED).contentType(MediaType.APPLICATION_JSON).body(resp.toString())
        } catch (e: java.lang.Exception) {
            println("Exception : " + e.message)
            ResponseEntity.status(HttpStatus.BAD_REQUEST).body(e.message)
        }
    }


    fun craftNotice(p: String, t: String, gameId: UniqueIdentifier): String {
        return "Hello, $p!\nYour super secret santa target for game # $gameId is $t."
    }

    fun sendEmail(recipient: String?, msg: String?): Boolean {

        // TODO replace the below line with your *REAL* api key!
        val SENDGRID_API_KEY = "SG.xxxxxx_xxxxxxxxxxxxxxx.xxxxxxxxxxxxxxxx_xxxxxxxxxxxxxxxxxx_xxxxxxx"
        if (SENDGRID_API_KEY == "SG.xxxxxx_xxxxxxxxxxxxxxx.xxxxxxxxxxxxxxxx_xxxxxxxxxxxxxxxxxx_xxxxxxx") {
            println("You haven't added your sendgrid api key!")
            return true
        }

        // you'll need to specify your sendgrid verified sender identity for this to mail out.
        val from = Email("test@example.com")
        val subject = "Secret Assignment from the Elves"
        val to = Email(recipient)
        val content = Content("text/plain", msg)
        val mail = Mail(from, subject, to, content)
        val sg = SendGrid(SENDGRID_API_KEY)
        val request = Request()
        return try {
            request.method = Method.POST
            request.endpoint = "mail/send"
            request.body = mail.build()
            val response = sg.api(request)
            println(response.statusCode)
            println(response.body)
            println(response.headers)
            true
        } catch (ex: IOException) {
            println(ex.toString())
            false
        }
    }


}
